Pass correct filename to relative require expressions

rktjmp opened the pull request:
When using "relative requires", the module name and module filename are passed into the code chunk when evaluating the "mod-name expression".

Currently this always uses the filename of the first expression, which ends up being `src/fennel/macros.fnl`. We can get the correct filename (the filename for the file that actually contains the "mod-name expression") from the next part of the AST.

It's probably pretty rare that someone wants to operate on the *filename* given inside the expression of a require, but it may as well be correct.

rktjmp commented on src/fennel/specials.fnl:
```diff
@@ -1149,12 +1149,21 @@ modules in the compiler environment."
+  ;; only gave a literal string the filename will be nil, but as that
+  ;; expression is static, that will be inconsequential.
+  (let [{2 second} ast
+        filename (or (and (utils.list? second) second.filename) :unknown)
```
Changed from `table?` to `list?` as the argument will always be a list? At least, it never seems to be a plain table.

rktjmp commented on src/fennel/specials.fnl:
```diff
@@ -1149,12 +1149,21 @@ modules in the compiler environment."
+  ;; `import-macros`, which wont have an appropriate filename to give to the
+  ;; lua chunk.
+  ;;
+  ;; We instead pull the filename from the user-suppied expression. If the user
+  ;; only gave a literal string the filename will be nil, but as that
+  ;; expression is static, that will be inconsequential.
```
What I mean here if unclear:

```
import-macros {...} :macros
```

`filename` will be nil, but `:macros` never attempts to use the value so it wont matter.

```
import-macros {...} (.. ... + :x)
```

`filename` will be correct here, and the user could destructure `...` into `modname filename`.

MuhammedZakir replied:
Typo: "user-suppied" -- "suppied" -> "supplied"

technomancy commented:
This looks sensible, and I think it will get us the desired effect.

However, it seems to be working around a weird implementation detail of `import-macros`. What if instead of working around this quirk, we changed `import-macros` to propagate the filename from the original AST rather than discarding it?

technomancy commented on src/fennel/specials.fnl:
```diff
@@ -1149,12 +1149,21 @@ modules in the compiler environment."
+  ;; We instead pull the filename from the user supplied expression. If the
+  ;; user only gave a literal string the filename will be nil, but as that
+  ;; expression is static, that will be inconsequential.
+  (let [{2 second} ast
```
This style of destructuring was used previously because we were pulling out numeric and string keys from a single table. However, in this case it would be better to use `[_ second] ast` instead since we got rid of the string key.

rktjmp commented:
> However, it seems to be working around a weird implementation detail of import-macros. What if instead of working around this quirk, we changed import-macros to propagate the filename from the original AST rather than discarding it?

Much smarter, updated.

technomancy commented:
Thanks; I've merged this!
